Populate Styleable entries correctly in R.java generated by MiniAapt

raviagarwal7 opened the pull request:
- Styleable entries array now contain right int values equal to its correspondding attr values. Android attributes are still given random values since R.java is not available at MergeAndroidResourcesStep for libraries.
- Set R idValues of a parent library equal to it’s dependent libraries idValue.

facebook-github-bot-5 commented:
By analyzing the blame information on this pull request, we identified @russellporter, @kageiit and @polycoder to be potential reviewers.

facebook-github-bot-5 commented:
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

facebook-github-bot-7 commented:
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

facebook-github-bot-3 commented:
@raviagarwal7 updated the pull request - [view changes](https://github.com/facebook/buck/pull/879/files/8746bc4464bf28c84e37b1778cf93fe71e8dfc36..872d4c7777becd7b90d7597692f410e8b6207226)

facebook-github-bot-7 commented:
@raviagarwal7 updated the pull request - [view changes](https://github.com/facebook/buck/pull/879/files/872d4c7777becd7b90d7597692f410e8b6207226..e72bab08bd1e5af03140c57bb71ca68ff6c66548)

raviagarwal7 commented:
@marcinkosiba, @russellporter -- bump for review

marcinkosiba commented on src/com/facebook/buck/android/MergeAndroidResourcesStep.java:
```diff
@@ -346,26 +351,77 @@ void writePerPackageRDotJava(
+              List<String> styleableIdValues = new ArrayList<>();
+              int styleableIndex = 0;
+
+              while (styleableIndex + index < linesInSymbolsFile.size()) {
```
not thrilled at the O(N^2) behavior here. I would prefer we run a single scan (you could just split on the first `_` and stick the results in a MultiMap)

raviagarwal7 replied:
this is still O(N). the outer loop goes from 0 to linesInSymbolsFile.size() & the inner loop goes through styleable int entries once more which is at-most N.

will move this to a multimap.

marcinkosiba commented on src/com/facebook/buck/android/aapt/RDotTxtEntry.java:
```diff
@@ -35,7 +35,7 @@
 
   // Taken from http://developer.android.com/reference/android/R.html
-  public static enum RType {
+  public enum RType {
```
why remove static?

raviagarwal7 replied:
removed as Intellij was showing a suggestion that nested enums are implicitly static - http://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.9

marcinkosiba commented on src/com/facebook/buck/android/MergeAndroidResourcesStep.java:
```diff
@@ -346,26 +351,77 @@ void writePerPackageRDotJava(
+      int index = 0;
+      while (index < linesInSymbolsFile.size()) {
+        RDotTxtEntry resource = getResourceAtIndex(linesInSymbolsFile, index);
+        index++;
```
your reuse of the index as the offset of the next resource later in the code is a bit confusing. Use a regular for loop and assign index +1 to an appropriately named variable.

marcinkosiba commented on src/com/facebook/buck/android/MergeAndroidResourcesStep.java:
```diff
@@ -346,26 +351,77 @@ void writePerPackageRDotJava(
+        } else {
+          if (resourceToIdValuesMap.get(resource) != null) {
+            resource = resource.copyWithNewIdValue(resourceToIdValuesMap.get(resource));
+          } else {
```
It would be great if the logic for the styleables lived in a separate method.

raviagarwal7 commented on src/com/facebook/buck/android/MergeAndroidResourcesStep.java:
```diff
@@ -346,26 +351,77 @@ void writePerPackageRDotJava(
+
+                if (styleableResource.idType == IdType.INT &&
+                    styleableResource.type == RType.STYLEABLE &&
+                    styleableResource.name.startsWith(resource.name)) {
```
good idea, will do

marcinkosiba commented:
@raviagarwal7, the comment from our bot points to a commit that doesn't seem to exist. Could you please confirm whether you updated this pull request and maybe try re-pushing the branch or something if you did?

raviagarwal7 commented:
Seems to be a bug in facebook bot while reporting changes ([view changes](https://github.com/facebook/buck/pull/879/files/e0631d1d685640d7571cf295ff361031d3bd0ced..69c000ce477bfeca596374cae32eb25da9c6cb80)) if the feature branch is rebased on master instead of merged.

I have squashed the commit with changes onto this branch - https://github.com/raviagarwal7/buck/commits/ravi/resource which is also rebased on top of master.

Here is full changes of the pull request - https://github.com/facebook/buck/pull/879/commits/69c000ce477bfeca596374cae32eb25da9c6cb80

marcinkosiba commented:
Thanks! It doesn't look like this bit about `getStyleableResources` was addressed:

> I think this is brittle - the code in the Merge step relies on the sorting behavior. We could add tests to make sure these don't get broken (yea, btw, add tests), but it seems to me a simpler approach would be to take all int styleable, split on the first _ and stick them in a multimap.

raviagarwal7 commented:
@marcinkosiba, using multimap to store the styleable ints now - https://github.com/facebook/buck/pull/879/commits/7f99ffcec494e47e607979acd23c9f0d61302c45. please re-review

marcinkosiba commented:
travis found a bunch of pmd errors, please address those.

marcinkosiba commented on src/com/facebook/buck/android/MergeAndroidResourcesStep.java:
```diff
@@ -345,26 +351,62 @@ void writePerPackageRDotJava(
+        resources.add(resource);
+        if (resource.idType == IdType.INT && resource.type == RType.STYLEABLE) {
+          String attrName = resource.name.substring(0, resource.name.indexOf('_'));
+          styleableIntArrToIntsMap.put(attrName, resource);
```
does it make sense to check that the key is not already present? I believe ImmutableSortedMap.Builder does this "for free"

raviagarwal7 replied:
styleableIntArrToIntsMap is a multimap so not needed, we expect multiple values.

marcinkosiba commented on src/com/facebook/buck/android/MergeAndroidResourcesStep.java:
```diff
@@ -345,26 +351,62 @@ void writePerPackageRDotJava(
-          resource = resource.copyWithNewIdValue(String.format("0x%08x", enumerator.next()));
+        } else {
+
+          if (resourceToIdValuesMap.get(resource) != null) {
```
could we get rid of one level of if nesting here by merging this into the else clause above?

raviagarwal7 replied:
done

raviagarwal7 commented:
@marcinkosiba just found out that using multimap and splitting on the first `_` is not feasible
```
int[] styleable PatternPathMotion { 0x7f060003 }
int styleable PatternPathMotion_patternPathData 0
int[] styleable PercentLayout_Layout {  }
int styleable PercentLayout_Layout_layout_aspectRatio 9
int styleable PercentLayout_Layout_layout_heightPercent 1
int styleable PercentLayout_Layout_layout_marginBottomPercent 6
int styleable PercentLayout_Layout_layout_marginEndPercent 8
```

we can have styleable int[] which have `_` in it's name eg: `PercentLayout_Layout`.
if we split on first `_` will result in no styleable int being found for `PercentLayout_Layout`.

I am going to add this testcase and update the pull request to rely on sorting.

raviagarwal7 commented:
@marcinkosiba, let me know about the next steps.

marcinkosiba commented:
For some reason the patch for the diff doesn't apply cleanly. I finally had time to look into it today, unfortunately my work was rewarded with a ton of failures around line length and checkstyle.
Please fix your IntelliJ line length/style settings. We have a correct project config checked into buck, not sure why your editor is not picking that up.

tl;dr; `ant lint` is failing on the diff.

raviagarwal7 commented:
thanks @marcinkosiba, for some reason intellij didn't pick up the project settings. anyways fixed the lint errors and rebased to master as well.

marcinkosiba commented:
this breaks a ton of our robolectric tests with errors similar to the one below:
```
java.lang.ArrayIndexOutOfBoundsException: 799060974
at android.content.res.TypedArray.getResourceId(TypedArray.java:570)
```

raviagarwal7 commented:
do you have a sample test case where it fails, that I can try out? otherwise it would be hard to figure out what's going on. It works well with our robolectric tests.

marcinkosiba commented:
> do you have a sample test case where it fails, that I can try out?

Not one easily available. I'll try and make time for this, but no promises. cc @dreiss in case he or his team can help out.

raviagarwal7 commented:
I am breaking down the pull request into smaller ones so it's easier to figure out what is causing the failures.

created the first one - https://github.com/facebook/buck/pull/1068

raviagarwal7 commented:
@dreiss -- I have rebased the pull request, could you please try importing and see if this works?

kageiit commented:
@dreiss any update on this?

dreiss commented:
Sorry for the delay.  It looks like the patch failed to apply.  Can you rebase and update?

raviagarwal7 commented:
@dreiss -- updated the pull request.

dreiss commented:
This had to be reverted due to a test failure. :(  I'll revive it and get more details.

kageiit commented:
@dreiss any update on the revert?

dreiss commented:
Yeah, this is definitely breaking some of our tests.  Some people who understand resources better than I do are trying to figure out exactly why.  It looks something like "now that the styleables are reporting proper resource ids, we're trying to load the resources (fonts) and failing because our resource path is not properly set."  I'll ping them.
